ENH: add time-window capability to .rolling

jreback opened the pull request:
xref #13327
closes #936 

This [notebook shows the usecase](http://nbviewer.jupyter.org/gist/jreback/186d09a99902a17a095d99ac6a5e4cd3)
- implement lint checking for cython (currently only for windows.pyx), xref #12995

This implements time-ware windows, IOW, to a `.rolling()` you can now pass a ragged / sparse timeseries and have it work with an offset (e.g. like `2s`). Previously you _could_ achieve these results by resampling first, then using an integer period (though you would have to jump thru hoops when crossing day boundaries and such). 

This now provides a nice easy / performant implementation (as indicated on the linked notebook, min/max impl is giving scaling issues).

```
In [1]: df = DataFrame({'B': range(5)})

In [2]: df.index = [Timestamp('20130101 09:00:00'),
   ...:             Timestamp('20130101 09:00:02'),
   ...:             Timestamp('20130101 09:00:03'),
   ...:             Timestamp('20130101 09:00:05'),
   ...:             Timestamp('20130101 09:00:06')]

In [3]: df
Out[3]: 
                     B
2013-01-01 09:00:00  0
2013-01-01 09:00:02  1
2013-01-01 09:00:03  2
2013-01-01 09:00:05  3
2013-01-01 09:00:06  4

In [4]: df.rolling(2, min_periods=1).sum()
Out[4]: 
                       B
2013-01-01 09:00:00  0.0
2013-01-01 09:00:02  1.0
2013-01-01 09:00:03  3.0
2013-01-01 09:00:05  5.0
2013-01-01 09:00:06  7.0

In [5]: df.rolling('2s', min_periods=1).sum()
Out[5]: 
                       B
2013-01-01 09:00:00  0.0
2013-01-01 09:00:02  1.0
2013-01-01 09:00:03  3.0
2013-01-01 09:00:05  3.0
2013-01-01 09:00:06  7.0
```

jreback committed:
ENH: add time-window capability to .rolling

xref #13327
CLN: pep for cython, xref #12995

jreback commented:
cc @jorisvandenbossche 
cc @wesm 
cc @chrisaycock 
cc @joshuastorck 

if anyone would like to try to fixup min/max would be great.

MaximilianR commented:
closes https://github.com/pydata/pandas/issues/936

jreback commented:
yeah we'll close #936 but I think that impl is a bit different.

chrisaycock commented:
Looks good to me. Regarding max/min, is the priority list faster than just recomputing the maximum value from scratch for each iteration? Caching and branch prediction can work wonders for small sample sizes.

wesm commented:
Long awaited feature! Will try to review this today.

joshuastorck commented:
Take a look at this: https://people.cs.uct.ac.za/~ksmith/articles/sliding_window_minimum.html.

You don't want to use a heap for min/max as it's worst case O(n * log(n)). There is a more efficient algorithm that is O(N). It also should be a bit easier to implement as all you really need for a data structure is a deque.

jreback commented:
@chrisaycock 

ok, silly of me not to test the completely naive soln, which is quite fast. It *still* scales on O(n) where n is the window size, but now the constant is 1/10 of my prioritylist constant. So this is pretty reasonable.

```
In [1]: N = 100000                     
dfp = DataFrame({'B' : np.random.randn(N)}, 
                 index=pd.date_range('20130101',periods=N,freq='s'))
   ...: 

In [2]: %timeit dfp.rolling(10).min()
100 loops, best of 3: 2.47 ms per loop

In [3]: %timeit dfp.rolling('10s').min()
100 loops, best of 3: 3.2 ms per loop

In [4]: %timeit dfp.rolling(100).min()
100 loops, best of 3: 2.58 ms per loop

In [5]: %timeit dfp.rolling('100s').min()
100 loops, best of 3: 11.2 ms per loop

In [6]: %timeit dfp.rolling(1000).min()
100 loops, best of 3: 2.62 ms per loop

In [7]: %timeit dfp.rolling('1000s').min()
10 loops, best of 3: 97.2 ms per loop
```

jreback commented:
@joshuastorck yes the deque is theoretically faster, but it was very awkward with a variable sized window; others welcome to take a shot thought!

wesm commented:
Meta-comment: since getting exposed to Gerrit over the last couple years, large reviews on GitHub PR's tend to encourage less careful review (because it's difficult to keep track of what still needs doing sometimes), at least for me. It might be worth playing around with gerrithub.io sometime to mirror patches there for review sometimes. Not sure what's the best solution.

wesm commented on pandas/core/window.py:
```diff
@@ -143,6 +171,21 @@ def __unicode__(self):
+            if index is None:
+                index = self._on
+            return index, index.asi8, self.win_unit
+        return index, index, self.win_unit
```
This is a little bit opaque. Is the purpose of the 2nd entry in the tuple to ensure you are always passing integers to the Cython function?

wesm commented on pandas/core/window.py:
```diff
@@ -847,10 +973,72 @@ class Rolling(_Rolling_and_Expanding):
+        elif (isinstance(self.obj, ABCDataFrame) and
+              on in self.obj.columns):
+            self._on = pd.Index(self.obj[on])
+            self.obj = self.obj.drop(on, axis=1)
```
Is there a better way to handle the "nuisance" columns in the common case where `on` is a timestamp column (or are there cases where this will not trigger a dataset copy)? In the future if we're able to do more zero-copy views (i.e. redesigned internals) then this concern goes away

jreback replied:
i'll have to look at this in more detail. short answer is no, but this *could* also be a view and not actually a copy, depends on numpy. (depends on whether can be sliced out vs internal take)

wesm commented on pandas/core/window.py:
```diff
@@ -847,10 +973,72 @@ class Rolling(_Rolling_and_Expanding):
+                freq = to_offset(self.window)
+            except (TypeError, ValueError):
+                raise ValueError("window string type in not "
+                                 "compat with a datetimelike index")
```
Include `self.window` in the error message?

jreback replied:
done

wesm commented on pandas/tests/test_window.py:
```diff
@@ -3047,3 +3072,470 @@ def test_expanding_apply(self):
+        result = df.rolling(window='1s', min_periods=1).sum()
+        expected = df.copy()
+        expected['B'] = [0.0, 1, 2, 3, 4]
+        tm.assert_frame_equal(result, expected)
```
From first principles, it's somewhat counter-intuitive to me that the result of a 1-second window on a ragged time series is the same size the original data-frame. The expectation, then, is that the user is always responsible for conforming the time series if that's what's desired?

wesm replied:
Scratch my comment, I misunderstood: the window is "include all data within the indicated time window of the given data point"

wesm commented on pandas/window.pyx:
```diff
@@ -96,757 +121,1281 @@ def _check_minp(win, minp, N, floor=1):
+        self.N = len(input)
+        self.minp = _check_minp(win, minp, self.N, unit=unit, floor=floor)
+        self.start = np.empty(0, dtype=int)
+        self.end = np.empty(0, dtype=int)
```
Can this cause problems on Win32 (or 32-bit platforms), where int is 32-bit?

jreback replied:
have some issues on this ATM

wesm commented on pandas/window.pyx:
```diff
@@ -96,757 +121,1281 @@ def _check_minp(win, minp, N, floor=1):
+
+
 @cython.boundscheck(False)
 @cython.wraparound(False)
```
Any reason not to disable these globally at the Cython module level?

`# cython: boundscheck=False`

jreback replied:
no, done

wesm commented:
This looks like it was quite a slog -- we should add to our ever-increasing TODO list a review of all of our aggregation code and ways to foster reuse or more flexibility. Trying to think if there was a way that this could have been made less painful. C++ templates could help (for example, a particular aggregator could be passed as a template argument, so the update inner loop could all get inlined at compile time -- e.g. the add_sum / remove_sum bits). But we'd need to decide if we want to commit to the route of writing more C++ and less Cython (I think there's good arguments for it, but evolving the build-and-deployment tools to accommodate that would be as discussed an upfront investment of work) -- this is somewhat orthogonal to the bigger discussion of writing more of the data structure internals in C or C++.

wesm commented:
Also, I'm always happy to see 4+ year-old issues that were once just a twinkle in my eye get realized, so better late than never =) Great stuff

joshuastorck commented:
I think there are a couple of compelling arguments for going to C++ templates, mostly that you can vary the implementation based on type at compile time. For example, since integer columns don't have NaN's, you don't really need the min_periods for rolling_* functions and could use SFINAE for that distinction. Alternatively, you could put 'if' statements inside the inner loops that are based on type, such as 'if std::is_integral<T>::value". With the rolling functions, much of the looping code is common, but the algorithm and state varies. You can define a rolling_apply function that takes iterators, a functor/lambda and variadic arguments to hold the state. And it goes without saying that there are a lot of possibilities for using SIMD when dealing with contiguous columns.

The downside with C++ is that you have to do a lot of boilerplate dispatching code to templates based on numpy type, the cross-platform compile gets more complicated, and the code could arguably be harder to read. 

At a minimum, it would be worth comparing the runtime of hand converted Cython code, as the code generator for Cython introduces a lot of constructs that might make it harder for the compiler to do optimizations, such as sanity checks in inner loops that could probably be done up front.

jorisvandenbossche commented:
@jreback There is a memory leak when using `center=True` (first time  I got a MemoryError, but second time it just crashed my computer :-))

jreback commented:
@jorisvandenbossche I didn't change anything w.r.t. ``center`` (and its disabled for rolling with time series). can you show a repro?

jorisvandenbossche commented:
With the example from your notebook: `df.rolling('2s', center=True).sum()`

jreback commented:
@jorisvandenbossche pushed. Its not implemented.

jorisvandenbossche commented on pandas/core/window.py:
```diff
@@ -268,20 +313,20 @@ class Window(_Window):
     Parameters
     ----------
-    window : int
+    window: int
```
@jreback It's the other way around (it needs a space, so these were good)

jreback replied:
ok......will change back

jorisvandenbossche commented on pandas/core/window.py:
```diff
@@ -821,23 +877,92 @@ class Rolling(_Rolling_and_Expanding):
+    4  4.0
+
+    A ragged (meaning not-a-regular frequency), time-indexed DataFrame
+    >>> df = DataFrame({'B': [0, 1, 2, np.nan, 4]},
```
Needs a blank line between the explanation and code block

jorisvandenbossche replied:
already did this in the PR

jreback replied:
yep, merged in

jorisvandenbossche commented on pandas/core/window.py:
```diff
@@ -821,23 +877,92 @@ class Rolling(_Rolling_and_Expanding):
+    Examples
+    --------
+
+    >>> df = DataFrame({'B': [0, 1, 2, np.nan, 4]})
```
Can you use `pd.` here in the docstring examples as well? (like you did in the whatsnew tutorial docs)

jorisvandenbossche replied:
already did this in the PR

jreback replied:
gr8!

jorisvandenbossche commented on pandas/core/window.py:
```diff
@@ -847,10 +972,77 @@ class Rolling(_Rolling_and_Expanding):
+            # must be monotonic for on
+            if not self._on.is_monotonic:
+                formatted = on or 'index'
+                raise ValueError("The on specified: {0} must be "
```
"the on specified" looks a bit strange here (certainly if you didn't specify it). Maybe just "{on} must be monotonic"?

jreback replied:
done

jorisvandenbossche commented on pandas/tests/test_window.py:
```diff
@@ -319,6 +319,13 @@ class TestRolling(Base):
+        df = DataFrame({'B': [0, 1, 2, np.nan, 4]})
+        df
+        df.rolling(2).sum()
+        df.rolling(2, min_periods=1).sum()
```
What is the purpose of this test? (the test name is also a bit strange)

jreback replied:
really just to make sure the doc-string actually executes

jorisvandenbossche commented:
@jreback Took a first look, I also made a PR with some small doc fixes that I made along the way: https://github.com/jreback/pandas/pull/17

Some other comments:

- If the `on` column is set anyway as the index in the result, I don't see that much of advantage of adding this keyword, as it is exactly the same as just doing a `set_index('foo').rolling(..)` ?

Further, I think this keyword only makes sense for datetimelike values with offset window? Probably we should raise in other cases that this is ignored? (if we keep the keyword)

- Can you update the docstring for `window` and `min_periods` keywords?

chrisaycock commented:
@jorisvandenbossche Indices are a huge nuisance for most users, so I had specifically requested the `on` keyword a while back.

jorisvandenbossche commented:
@chrisaycock should the resulting frame then keep it as a column? (now the 'on' column is set as the index, so if you want it as a column, you have to do `reset_index()` after rolling)

jreback commented:
@jorisvandenbossche actually that last is an error

```
In [1]:    dft = DataFrame({'B': [0, 1, 2, np.nan, 4]},
   ...:                    index = pd.Index([pd.Timestamp('20130101 09:00:00'),
   ...:                                      pd.Timestamp('20130101 09:00:02'),
   ...:                                      pd.Timestamp('20130101 09:00:03'),
   ...:                                      pd.Timestamp('20130101 09:00:05'),
   ...:                                      pd.Timestamp('20130101 09:00:06')],
   ...:                                     name='foo'))

In [2]: dft
Out[2]: 
                       B
foo                     
2013-01-01 09:00:00  0.0
2013-01-01 09:00:02  1.0
2013-01-01 09:00:03  2.0
2013-01-01 09:00:05  NaN
2013-01-01 09:00:06  4.0

In [3]: df = dft.reset_index()

In [4]: df
Out[4]: 
                  foo    B
0 2013-01-01 09:00:00  0.0
1 2013-01-01 09:00:02  1.0
2 2013-01-01 09:00:03  2.0
3 2013-01-01 09:00:05  NaN
4 2013-01-01 09:00:06  4.0

#### this should *not* set the index, instead it should remain a column (as it was to start), and the existing index should be preserved.
In [5]: df.rolling('2s', on='foo').sum()
Out[5]: 
                       B
foo                     
2013-01-01 09:00:00  0.0
2013-01-01 09:00:02  1.0
2013-01-01 09:00:03  3.0
2013-01-01 09:00:05  NaN
2013-01-01 09:00:06  4.0
```

jreback commented:
ok, fixed up docs as suggested by @jorisvandenbossche 

also, ``on`` now works correctly in that it doesn't actually set the index; further its now cheap
to use ``on`` as just skipped during block creatino and added back at the end. Further now the
result frame will be in the same order as input. 

Using a slightly different example for testing.
```
In [1]:        df = DataFrame({'A': [0, 1, 2, 3, 4],
   ...:                         'B': [0, 1, 2, np.nan, 4],
   ...:                         'C': pd.Index([pd.Timestamp('20130101 09:00:00'),
   ...:                                        pd.Timestamp('20130101 09:00:02'),
   ...:                                        pd.Timestamp('20130101 09:00:03'),
   ...:                                        pd.Timestamp('20130101 09:00:05'),
   ...:                                        pd.Timestamp('20130101 09:00:06')])},
   ...:                        columns=['A', 'C', 'B'])

In [2]: df
Out[2]: 
   A                   C    B
0  0 2013-01-01 09:00:00  0.0
1  1 2013-01-01 09:00:02  1.0
2  2 2013-01-01 09:00:03  2.0
3  3 2013-01-01 09:00:05  NaN
4  4 2013-01-01 09:00:06  4.0

In [3]: df.rolling('2s', on='C').sum()
Out[3]: 
     A                   C    B
0  0.0 2013-01-01 09:00:00  0.0
1  1.0 2013-01-01 09:00:02  1.0
2  3.0 2013-01-01 09:00:03  3.0
3  3.0 2013-01-01 09:00:05  NaN
4  7.0 2013-01-01 09:00:06  4.0

In [4]: df.rolling('2s', on='C').B.sum()
Out[4]: 
0    0.0
1    1.0
2    3.0
3    NaN
4    4.0
Name: B, dtype: float64

In [5]: df.rolling('2s', on='C')[['A', 'B', 'C']].sum()
Out[5]: 
     A    B                   C
0  0.0  0.0 2013-01-01 09:00:00
1  1.0  1.0 2013-01-01 09:00:02
2  3.0  3.0 2013-01-01 09:00:03
3  3.0  NaN 2013-01-01 09:00:05
4  7.0  4.0 2013-01-01 09:00:06
```

jreback commented:
further comments?

@jorisvandenbossche 
@wesm 
@chrisaycock

wesm commented on doc/source/computation.rst:
```diff
@@ -391,6 +391,67 @@ For some windowing functions, additional parameters must be specified:
+
+   dft.rolling('2s').sum()
+
+Using a non-regular, but still monotonic index, rolling with an integer window does not impart any special calculation.
```
Code examples aside, from this description it isn't perfect clear what the time window does, precisely. Essentially the job is to distinguish this clearly from `resample` (which has its own complexities with anchored vs non-anchored time offsets). 

Perhaps something like: "For each time point, includes all preceding values occurring within the indicated time delta." If it confuses people we can always create a diagram (for example: I will probably include this in the 2nd ed of my book and almost certainly create a diagram to make it clear).

wesm commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -111,6 +112,64 @@ that forward filling happens automatically taking the most recent non-NaN value.
+
+   dft = dft.reset_index()
+   dft
+   dft.rolling('2s', on='foo').sum()
```
This content is duplicated from the other rst file, maybe this could be `include`-ed from a shared file?

wesm replied:
Or simply link to the other file

jreback replied:
well, have been doing this for quite a while, showing a complete example in whatsnew is pretty helpful, yes its duplicated, but an in-line example is pretty nice (rather than just a link).

wesm commented:
I think this is looking pretty good. Are you satisfied with the unit tests for correctness? If I'm being perfectly honest it feels a touch thin -- would it make sense to run a test like https://github.com/pydata/pandas/pull/13513/files#diff-51f469e6b0911fa166d4cc24e5123919R3513 on maybe a slightly smaller dataframe size but on all the rolling statistics? Maybe I missed some other tests that do that

chrisaycock commented:
All looks good to me. I look forward to getting this into people's hands.

shoyer commented:
This functionality looks great, but I worry about overloading the first argument to `.rolling` with this orthogonal functionality (frequencies/label-units as well as integers). This is similar to the issues with fallback logic for ix/getitem. If the index is numeric, it will no longer be entirely obvious what `df.rolling(2)` means.

jreback commented:
@shoyer this is not the same as fallback logic at all as its an explicit user choice. If you pass an offset and you have a time-based on/index then it will work, otherwise it will raise. Adding *another* method to basically to the same ``.rolling_time`` just crowds the api and then forces users to jump thru too many hoops (what if I remove the index and switch it to a column, now I *also* have to change the method name....bad).

I *agree* however, that we could *deprecate* using integer based with time-based on/index, though It IS still meaningful with a non-ragged freq. Up till now that has been the only way.

jreback commented:
@wesm https://github.com/pydata/pandas/pull/13513/files#diff-51f469e6b0911fa166d4cc24e5123919R3568 covers the complete comparison. Though its a bit simplistic. I will add a more complete ragged example.

jreback commented:
[here](https://github.com/pydata/pandas/pull/13513/commits/c9c420268af2748c88a8f708366bd08155eb4dea) is a doc update, is this more clear?

@wesm

jorisvandenbossche commented:
> >  Stephan:  I worry about overloading the first argument to .rolling with this orthogonal functionality

> Jeff: Adding another method to basically to the same .rolling_time just crowds the api

@jreback It does not necessarily have to be another method, we could also use another *keyword* to not overload the `windows` keyword.

jreback commented:
more [testing](https://github.com/pydata/pandas/pull/13513/commits/7c17e0bb5b19947305d73a6e1a5330bae0ff7432)

jreback commented:
> @jreback It does not necessarily have to be another method, we could also use another keyword to not overload the windows keyword.

again I disagree, this is just more confusion. ``window`` is the obvious keyword here. This an unambiguous operation, no inference at all. It either has a compat index or it raises. 

We maybe *could* have used ``freq`` (which we deprecated in 0.18.0), but then this gives more confusion IMHO.

```
df.rolling(window='3s')
df.rolling(freq='3s')
```

not really clear what ``freq`` means here.

jorisvandenbossche commented:
The difference is that up to now,  `window` expressed a "fixed number of rows", while in this PR we add a way of specifying the window as a "fixed interval" (with possibly varying number of rows). 

I think the possible confusion that @shoyer points to (correct me if I am wrong!), is that `windows=2` could in this light also be interpreted as a fixed interval window instead of fixed number of rows window (as in principle, nothing prevents to also have this functionality for numeric indices, although it is now only for datetime). So in that sense two separate keywords for the two different notions of windows would make sense IMO.

But I have to say, I have not directly an idea for a good (and short) keyword name for the fixed interval window.

shoyer commented:
> The difference is that up to now, window expressed a "fixed number of rows", while in this PR we add a way of specifying the window as a "fixed interval" (with possibly varying number of rows).
> 
> I think the possible confusion that @shoyer points to (correct me if I am wrong!), is that windows=2 could in this light also be interpreted as a fixed interval window instead of fixed number of rows window (as in principle, nothing prevents to also have this functionality for numeric indices, although it is now only for datetime). So in that sense two separate keywords for the two different notions of windows would make sense IMO.

@jorisvandenbossche Yes, this is my exact concern. Different keywords or methods names would resolve this ambiguity.

@jreback  I don't think it would be a good idea to switch the default for only time windows -- it's better to have methods that act the same on any dtype of index.

I agree that `window` and `freq` are bad names for expressing this, because it's not clear which one is for label width and which is for the number of rows.

Some possible names:
- for fixed interval windows: `interval` (this is all I have off the top of my head)
- for fixed number of rows: `count`, `size` (maybe a little ambiguous?), `n_rows` (maybe too limiting if you could do a rolling window over other axes?)

Here are two API options I like:
1. Rename `window` -> `count` and add another keyword argument `interval` to handle this new capability. So we would have `.rolling(count=2)` vs `.rolling(interval='2s')`. `count` would be the default, and you would get a `TypeError` if you attempt to set both `count` and `interval`.
2. Create two different methods, `rolling_count` and `rolling_interval`. Deprecate plain `rolling` (alias it to `rolling_count`). This is a little cleaner from an API design perspective because it avoids two mutually exclusive signatures on a single method.

chrisaycock commented:
Note that `.rolling()` isn't the only function. There's `.rolling_mean()`, `.rolling_min()`, `.rolling_sum()`, etc. If we split *count* and *interval* into different functions, we'll have two versions of **all** of these.

I quite like the notion of just one function and one parameter, where 
 - 2 means two indices, and
 - '2s' means two seconds in the index.

That seems pretty straightforward to me.

shoyer commented:
> Note that .rolling() isn't the only function. There's .rolling_mean(), .rolling_min(), .rolling_sum(), etc. If we split count and interval into different functions, we'll have two versions of all of these.

@chrisaycock This isn't quite right -- `rolling_mean()` and friends have been deprecated: http://pandas.pydata.org/pandas-docs/version/0.18.1/whatsnew.html#window-functions-are-now-methods

wesm commented:
@jreback the doc update looks good, thank you

jorisvandenbossche commented:
I find `window_size` and `window_interval` quite clear and descriptive names, but a bit long .. (so size/count and interval are maybe better)

At the moment I tend more toward the two keywords, and not two functions: it would be a bit annoying to already deprecate what was just introduced, and in particular `rolling_count()` looks to much on the current `pd.rolling_count()`/`rolling().count()` I think (and I don't think of better names).

chrisaycock commented:
@shoyer Ah, yes. I had completely forgotten about the deprecation.

jreback commented:
I still don't see any issue with this. if the window is an offset then the ``on`` must be a datetime, if its an integer then it doesn't matter. As I said above, I could see deprecating an integer for a datetime ``on`` (you don't really lose anything by thing, you can simply ``.reset_index()`` if you want). 

But this is completely unambiguous now.

What exactly is the problem? That one misinterprets an integer on a time-based index? well that is the case we have now.  ``window=2`` is completely unambiguous as is ``window='2s'``. If you don't have a time-based index, then ``window='2s'`` would raise.

Further I don't see *anything* other than a fixed number of rows in our current implementation (yes If we had ``Interval('[0, 5)'])`` which is essentially a Period for integers, then we could have variable aggreggation on integer indexes), BUT its IS STILL unambiguous.

An API should be simple, complete and unambiguous. I think adding different keywords just introduces much more confusion / complexity here.

shoyer commented:
@jreback It would also be nice to have numeric interval based rolling windows. In that case, `window=2` becomes ambiguous, because it's not clear if it means "span 2 rows" or "span an interval of width 2".

So you're right, that is currently unambiguous, but only if we take a short sided view that precludes any possibility of supporting this feature for non-time windows.

jreback commented:
@shoyer I still disagree.

you would simply specify ``window=Interval(0, 5, closed=True)`` for an integer based aggregator.

I think this is completely unambiguous and as a user you don't really have to think too hard.

e.g. should I use the ``window`` kw or the ``interval`` kw? you just select *what* you actually want to do.

much simpler.

shoyer commented:
> you would simply specify `window=Interval(0, 5, closed=True)` for an integer based aggregator.

@jreback It doesn't make sense to provide a fixed start and end for a rolling window interval, so the `Interval` class wouldn't work. If you want to be fully explicit, I suppose you could write something like `window=NumberDelta(5)`. But generally we just use numbers for the difference of two numbers. It would be pretty unnatural to create a custom NumberDelta class for this use.

jreback commented:
again I disagree. ``Interval(0, 5)`` is perfectly unambiguous. I want a width of 5 non-inclusive. It IS a number-delta. 

Of course we don't have ``Interval`` at all, so this is quite moot.

shoyer commented:
@jreback I think you're mixing up two uses of the word "interval" (which may indicate that this is a bad keyword argument name).

`Interval` is `Period` like with a start and an end. But `Period('2000-01-01')` doesn't make sense as an input to rolling -- we want something like an time-offset/time-delta for indicating window width, not a difference with a fixed position. `Interval(0, 5)` means the particular window from 0 to 5, not a generic window of width 5. 

In contrast, by `interval` I meant a simple difference between two numbers (used to label the values in the window).

jreback commented:
@shoyer not at all. I understand exactly what ``Interval(0, 5)`` means. And in fact I think it acts exactly like a number-delta. But in any event this doesn't exist and is moot. I think this is going far afield.

Since this is unambiguous, simple and complete. I still don't see any objections here. Or even a concrete example.

jorisvandenbossche commented:
> Interval(0, 5) -> in fact I think it acts exactly like a number-delta.

It *can* act like a number-delta by taking the length of it, but it is *not* a number-delta, as `Interval(5,10)` would mean the same number delta.

And anyway, if we want numeric-delta window functionality, a separate keyword for that (eg `window_delta=2`) seems much simpler to me than having to pass a specific object for it (eg `window=pd.NumberDelta(2)`)

I agree it is unambiguous in the current implementatin, meaning that code-wise we can perfectly distuinguish which of the two cases the user wants, but it are still two very different concepts (window as fixed number of rows vs window as fixed width with variable number of rows). IMO that indicates it is not a good idea to use the same keyword for it (although I understand that you then loose the convenience of not having to specify the keyword name, like `rolling('2s')` instead of `rolling(width='2s')`

joshuastorck commented:
I would agree that we want to use "delta" terminology and not "interval" terminology. The current behavior where you specify an integer is in units of "SampleDelta" or maybe "ILocDelta". Using NumberDelta and time deltas make perfect sense to me.

The C++ analog to this would be:

template <typename InputIterator, typename DeltaType = decltype(*declval<InputIterator>() - *declval<InputIterator>())>

wesm commented:
On the 1-argument-2-interpretations (fixed versus variable window) issue. I admit I've been the biggest offender in the argument overloading / polymorphism department. Overall, I think it can improve the ergonomics of using a particular method, but there can also be confusion or ambiguities. 

In this particular case I don't think it's too egregious. To satisfy explicitness, we might add keyword arguments `window_periods` or `window_interval` keyword arguments, then assign the first positional `window` argument to one of these based on the type (int, string, DateOffset, etc.).

so in summary, I'm +0 on the current proposal. It does introduce some inflexibility if we want to add "yet another window variety"

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -3,15 +3,16 @@
 v0.19.0 (August ??, 2016)
 -------------------------
 
-This is a major release from 0.18.2 and includes a small number of API changes, several new features,
```
@jorisvandenbossche slight changes need here (which are in master). done here i think is fine.

jreback commented:
ok, well I'd like to push this into master now; now that we are on 0.19, this can be tried out for a bit by the deer-in-the-headlights crowd who run pandas master 👯 

If there are later suggestions for a different API, then these can be addressed then. To be honest this is a highly theoretical use case where you actually want separate args. So I think that we'd need to go whole hog and deprecate ``window`` and add 2 new args instead.

But that is really -1 from me.

jreback commented:
any final comments? / merging soon.

shoyer commented:
Well, I'm still not happy with this API choice but I suppose I can live with @wesm's suggested compromise. It seems very awkward to me, though.

shoyer commented:
And to be clear, by "awkward" I mean that I think the current implementation will lead to bugs and confusion in user code. But that's the price we pay for convenience...

wesm commented:
@shoyer so to make sure I understand the core concern, it's that this overloads the semantics of the window argument -- either you get a fixed number of periods or a variable-length window based on the type of input? We could certainly introduce auxiliary APIs `rolling_fixed` and `rolling_variable` (leaving the existing `rolling`) which only accept one input or the other to offer a more precise / readable result. 

On the other hand, you could imagine having code like

```
def my_rolling_stat(df, win):
    return df.rolling(win).mean()
```

(this might be a lot more complex in practice), and this could be easily switched from variable window to fixed window through the `win` argument

wesm commented:
At least, I feel the current state of things is much better / more consistent than the earlier smorgasbord of `rolling_*` functions

shoyer commented:
> @shoyer so to make sure I understand the core concern, it's that this overloads the semantics of the window argument -- either you get a fixed number of periods or a variable-length window based on the type of input? 

Yes, this is my concern. I strongly prefer function arguments that only do one thing, independent of the type of the data -- especially when the behavior is potentially ambiguous, as in this case.

 > We could certainly introduce auxiliary APIs rolling_fixed and rolling_variable (leaving the existing rolling) which only accept one input or the other to offer a more precise / readable result.

Yes, this is another clean way to handle the API.

> On the other hand, you could imagine having code like
```
def my_rolling_stat(df, win):
    return df.rolling(win).mean()
```
> (this might be a lot more complex in practice), and this could be easily switched from variable window to fixed window through the win argument

Indeed, it's certainly more convenient -- at least at first -- to reuse a single argument for these two different purposes. This could even make sense for special purpose code, e.g., if the user is happy to make the assumption that `df` will always have a `DatetimeIndex`. But I don't think these sort of special rules make sense for pandas, which does an admirable job of providing a composable language for data analysis that works consistently independently of the particular data types. I also don't think it's hard for a user to add some custom logic before calling `.rolling()` to implement this sort of functionality themselves.

 > At least, I feel the current state of things is much better / more consistent than the earlier smorgasbord of rolling_* functions

Totally agreed! I'm very happy we were able to switch to the `.rolling()` methods.

chris-b1 commented:
This is smaller, but is `on=` the right keyword for column selection?  xref #13500 - I'd want to use the same keyword there. 

Although I suggested `key=` in that issue, I'm now thinking that `by=` would be better, for consistency with groupby and the sorting api?

jreback commented:
we use on for pd.merge already

jorisvandenbossche commented:
We are a bit stuck it seems ... Below I tried to list the possibilities to go forward based on the discussions above (note: all used keyword/method names are of course open for discussion :-), I used 'size' and 'delta' for now to make a distinction)

1. The current PR: `DataFrame.rolling(window, ...)`
  * Use the existing `window` argument for both types of window
  * Advantage: simple usage (`df.rolling('2s')`)
  * Disadvantage: overloading of one keyword (`window`)

2. Add separate keyword for new functionality: `DataFrame.rolling(window, window_delta=.., ...)`
  * Advantage: no overloading of the same keyword with different behaviours
  * Disadvantage: not being able to use the shorthand `df.rolling('2s')` for the new functionality, as you always have to use a named keyword: `df.rolling(window_delta='2s')`, so worse ergonomics
  * *Alternative 1*: introduce two new keywords: `DataFrame.rolling(window_size, window_delta, ...)` to have a more descriptive keyword for the existing functionality as well.
    * Note that there is not necessarily a need to deprecate the `window` keyword. Eg `DataFrame.rolling(window_size=None, window_delta=None, ..., **kwargs)` -> `window_size` can handle the the current use of positional arg, and the keyword use `window=` can be checked for in the kwargs.
  * *Alternative 2* (from @wesm): introduce two new keywords for explicitness, but keep current `window` for convenience: `DataFrame.rolling(window, window_size=None, window_delta=None, ...)`, and then possibly assign the first positional window argument to one of these based on the type (int, string, DateOffset, etc.).

3. Add separate method for new functionality: `DataFrame.rolling_delta(window, ...)`
  * Advantage: no overloading of the same keyword/function with different behaviours
  * Disadvantage: yet another method(s) added to the DataFrame namespace
  * Options: also add a `DataFrame.rolling_fixed()` (but keep existing `rolling` as an alias for this)

jreback commented:
I still don't buy that this overloading is bad in any way.

I'd like @shoyer to convince why having the API in the is PR is not a good idea.

> And to be clear, by "awkward" I mean that I think the current implementation will lead to bugs and confusion in user code. But that's the price we pay for convenience...

This is just bunk. it is more *awkward* to having confusing arguments in the first place which is what 2/3 by @jorisvandenbossche suggest.

Python is NOT c++, nor is it Java. We should be pythonic and follow current conventions.  Pandas DOES quite a lot of inference. The POINT is to have convenience.

Trying to change something just for the sake of some theoretically purity is not in the spirit of python (and pandas at all). In fact, I would argue that making this NOT similar to simple, unambiguous usage is a MAJOR break in and of itself. 

You have to have a VERY convincing argument to actually change things.

chrisaycock commented:
I agree with @jreback. Overloading the parameters makes things pretty easy since the user needs only to know about one input. Let the machine figure-out what to do with it.

shoyer commented:
> Python is NOT c++, nor is it Java. We should be pythonic and follow current conventions. Pandas DOES quite a lot of inference. The POINT is to have convenience.

There is certainly precedence for this sort of behavior in pandas, but it's not the good kind -- it's confusing features like `.ix` or the overloading of `__getitem__`. Things that require entries on the "gotchas" page.

Can you find a single example of this sort of thing in the Python standard library? In my view it is very non-Pythonic, in addition to being poor software design more generally. "In the face of ambiguity, refuse the temptation to guess."

chris-b1 commented:
What feels different about this one (vs. `.ix` - which I agree is too magical) is that there is no guessing or fallback - there are well defined behaviors or error messages on each path.

I do see the point about adding an interval window in the future...but even then - say this was the API - is the second row really ambiguous?
```python
df.rolling('2s').sum()
df.rolling(5).sum()
df.rolling(numeric_interval=5.0).sum()
```

wesm commented:
Here's my advice for how to move forward:

- Proceed for now with the overloaded argument. This is much lower risk for conflation than .ix had with labels vs integers

- Observe user behavior and see how it goes for people. 

- We always have the option to add explicit keyword arguments

- If the string argument is causing problems, we can decide to deprecate and encourage using explicit arguments. This is obviously a last resort.

While not exactly equivalent, I see this as similar to 'ndarray.astype' -- effectively what we are doing is coercing the argument to a Window spec. We might consider adding a "WindowSpec" type of object that can be passed. 

The variable windows will only be used by a pretty narrow subset of users -- I agree we should continue to be careful about argument overloading.

shoyer commented:
@chris-b1 I agree, this is not as bad as `.ix`. I still don't like functionality specific to the dtype of the function arguments.

@wesm OK, I can live with that. I'm pretty sure that we'll eventually need to add explicit keyword arguments, because there's no other sane way to handle fixed label width rolling windows for non-time labels.

`WindowSpec` is indeed one way to make this explicit, although it feels rather contrived to me -- it's a way to make an ambiguous API explicit without breaking it, not how you would design an explicit API from scratch.

jreback commented:
@shoyer I think you missed my original point. *pandas* does lots of 'magical' things. Changing something to non-magical will be a huge change. Yes you can attempt to forestall increasing the magic, but decreasing magic itself is a rather large API change.

You certainly could, but it should then be vetted like any other API change. 

merging

shoyer commented:
> Changing something to non-magical will be a huge change. Yes you can attempt to forestall increasing the magic, but decreasing magic itself is a rather large API change.

This is exactly why I argued for not adding the magic here in the first place....

jreback commented:
@shoyer 

> You certainly could, but it should then be vetted like any other API change.

pandas *already* has magic around these types of things. changing it is bad. (unless you basically change everything)

wesm commented:
hey folks, just a meta-comment on this discussion. While these are all plain text exchanges and tone is devilishly hard to convey, I encourage all to be particularly mindful of how this discussion might read to outside observers or would-be contributors to the project. 

I realize that the disagreement about the API dragged on for about 2 weeks, and I think we all would agree that 2 weeks is too long, and some frustration about this is understandable (always be closing, right?). Can we come up with a working plan for the future when resolving these sorts of contentious issues? I should have intervened sooner to resolve the impasse (bad timing: had a particularly busy couple of weeks, and also was on vacation and not looking at e-mail / github very much). The plan might be just to bug me to help arbitrate.

As far as the specifics of points of view were argued, as frustrating as it may be, there's some things we should definitely avoid. I hate to nail a particular comment to the wall but: https://github.com/pydata/pandas/pull/13513#issuecomment-233936362 — statements like "this is just bunk" are not helpful. It is OK to place the burden of proof on the person raising objections, but I think a more constructive way would be "I still disagree with you about this. Can you give some concrete examples illustrating why you feel this way?" We won't always agree but we should always strive to validate others' concerns and keep the conversation both constructive and productive (which I think we can do without being needling or dismissive).

I agree that pandas has teetered on the spectrum between preciseness / unambiguity and convenience, and we (e.g. yours truly) have made some lapses in judgment that have been costly to remedy (like `.ix` and the integer indexing problem). I try to avoid "Edge Case Driven Development", unless the edge cases can be demonstrated to be particularly harmful to the users (for example: anything that would cause a user to unwittingly obtain erroneous output with some reasonable probability). 

Thanks

magicmathmandarin commented:
I find the time-based window implementation to be confusing.  
For example, 
```
df = pd.DataFrame({'x': [0, 1, 2, np.nan, 4]},
                    index=pd.date_range('20200101',
                    periods=5, freq='d'))
df.rolling(2).sum()
```
and 
`df.rolling('2d').sum()`
return totally different results. 
 df.rolling('2d').sum() seems to assume min_periods =1.  
A user has to go through trial and error to figure that out, and cannot be sure whether the result would be as expected for a different case.

albertbuchard commented:
As @magicmathmandarin pointed out, min_periods does not work as expected, just try:

```
df.rolling("2s", min_periods=2).sum()

                                       B
2013-01-01 09:00:00  NaN
2013-01-01 09:00:02  NaN
2013-01-01 09:00:03  3.0
2013-01-01 09:00:05  NaN
2013-01-01 09:00:06  7.0
```
